### PR TITLE
B2MD: Don't try to reclaim if continuing to parse

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

B2MD called out to the decoder's `shouldReclaimBytes` after every
parsing attempt, even if the parser said `.continue`.

That's quite pointless because we won't add any bytes into the buffer
before we're trying the parser again.

Modifications:

Only ask the decoder if we should reclaim bytes if the decoder actually
says `.needMoreData`.

Result:

Faster, better, more sensible.

### CODE REVIEW

**Comment by weissi** [comment]
CC @fabianfett

File: Tests/NIOTests/CodecTest.swift [hunk 1]
```diff
@@ -200,22 +200,24 @@ public final class ByteToMessageDecoderTest: XCTestCase {
         XCTAssertEqual(testDecoderIsNotQuadratic_reallocs, 3)
     }
 
-    func testMemoryIsReclaimedIfMostIsConsumed() throws {
+    func testMemoryIsReclaimedIfMostIsConsumed() {
         let channel = EmbeddedChannel()
         defer {
-            XCTAssertNoThrow(try channel.finish())
+            XCTAssertNoThrow(XCTAssertTrue(try channel.finish().isClean))
```
Comment (fabianfett) [inline Tests/NIOTests/CodecTest.swift:206]: I think we can get rid of the enclosing `XCTAssertNoThrow`...

Comment (weissi) [reply Tests/NIOTests/CodecTest.swift:206]: @fabianfett but we want that?

Comment (weissi) [reply Tests/NIOTests/CodecTest.swift:206]: it has a `try` and I don't want it to throw, right?

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Review by Lukasa — APPROVED** [review]

File: Tests/NIOTests/CodecTest.swift (hunk 1 above)
Comment (fabianfett) [reply Tests/NIOTests/CodecTest.swift:206]: Correct, but AFAIK XCTAssertTrue also catches `try`.

Comment (fabianfett) [reply Tests/NIOTests/CodecTest.swift:206]: In other word `XCTAssertTrue` has a build in `XCTAssertNoThrow`

**Comment by weissi** [comment]
@swift-nio-bot test perf please

File: Tests/NIOTests/CodecTest.swift (hunk 1 above)
Comment (fabianfett) [reply Tests/NIOTests/CodecTest.swift:206]: Probably not available for Swift 5.0.

**Review by fabianfett — APPROVED** [review]
LGTM

**Comment by weissi** [comment]
> circular_buffer_into_byte_buffer_1kb	0.051517237	0.041180082	previous	25%
> circular_buffer_into_byte_buffer_1mb	0.094858812	0.082505024	previous	14%
> byte_buffer_view_iterator_1mb	0.117302428	0.126364362	current	-7%

@Lukasa / @tomerd I didn't touch anything that's related to these tests, yet we get _huge_ differences there. Does that make sense to anybody? (I did retry this already and the results before were similar)

**Comment by Lukasa** [comment]
Are we consistently getting huge differences?

**Comment by weissi** [comment]
> Are we consistently getting huge differences?

at least twice in a row... I'll try again

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@Lukasa safe to say that yes, in three consecutive runs, we're seeing >20% diffs here. And this PR only touches B2MD but the diffs are in unrelated tests...

**Comment by Lukasa** [comment]
Seems like we need to investigate these tests and the perf environment. It could be that there is wide variance in the tests (e.g. maybe the bulk of their runtime is in the allocator), or that there is wide variance in the test environment.

**Comment by weissi** [comment]
@Lukasa switching the investigation over to https://github.com/apple/swift-nio/pull/1716 which is an empty PR.

**Comment by weissi** [comment]
@Lukasa don't think it's the allocator or something because the results themselves are pretty stable (std diff 0.00024521498109819677)

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@Lukasa okay, something funny's going on with the binary here. Merely adding the benchmark (with no code in there) triggers the major diffs. Will need to check inlining thresholds and other code layout changes now. This'll be a while

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Comment by weissi** [comment]
@fabianfett / @Lukasa we can safely ignore the performance "regression". Details are the motivating example of this new doc: https://github.com/apple/swift-nio/pull/1738

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Review by Lukasa — APPROVED** [review]

**Pull request merged**